File: tennis/typescript/src/tennis-game.ts
export enum Player {
  One,
  Two,
}

export interface TennisGame {
  wonPoint(player: Player): void;
  getScore(): string;
}

const POINT_NAMES = ['Love', 'Fifteen', 'Thirty', 'Forty'];

export class Game implements TennisGame {
  private readonly points: [number, number] = [0, 0];

  wonPoint(player: Player): void {
    if (this.getScore().startsWith('Win for')) {
      throw new Error('The game has already been won');
    }

    this.points[player]++;
  }

  getScore(): string {
    const [pointsOne, pointsTwo] = this.points;

    if (pointsOne === pointsTwo) {
      return pointsOne < 3 ? `${POINT_NAMES[pointsOne]}-All` : 'Deuce';
    }

    if (pointsOne >= 4 || pointsTwo >= 4) {
      const leader = pointsOne > pointsTwo ? Player.One : Player.Two;

      return Math.abs(pointsOne - pointsTwo) === 1
        ? `Advantage ${this.playerName(leader)}`
        : `Win for ${this.playerName(leader)}`;
    }

    return `${POINT_NAMES[pointsOne]}-${POINT_NAMES[pointsTwo]}`;
  }

  private playerName(player: Player): string {
    return player === Player.One ? 'Player One' : 'Player Two';
  }
}

File: tennis/python/tennis_game.py
from enum import Enum, auto
from typing import Protocol


class Player(Enum):
    ONE = auto()
    TWO = auto()


class TennisGame(Protocol):
    def won_point(self, player: Player) -> None: ...

    def get_score(self) -> str: ...


_POINT_NAMES = ["Love", "Fifteen", "Thirty", "Forty"]

_PLAYER_NAMES = {
    Player.ONE: "Player One",
    Player.TWO: "Player Two",
}


class Game(TennisGame):
    def __init__(self):
        self._points = {Player.ONE: 0, Player.TWO: 0}

    def won_point(self, player: Player) -> None:
        if self.get_score().startswith("Win for"):
            raise RuntimeError("The game has already been won")

        self._points[player] += 1

    def get_score(self) -> str:
        points_one = self._points[Player.ONE]
        points_two = self._points[Player.TWO]

        if points_one == points_two:
            return f"{_POINT_NAMES[points_one]}-All" if points_one < 3 else "Deuce"

        if points_one >= 4 or points_two >= 4:
            leader = Player.ONE if points_one > points_two else Player.TWO

            return (
                f"Advantage {_PLAYER_NAMES[leader]}"
                if abs(points_one - points_two) == 1
                else f"Win for {_PLAYER_NAMES[leader]}"
            )

        return f"{_POINT_NAMES[points_one]}-{_POINT_NAMES[points_two]}"
